### PR TITLE
[ENH][SCHEMA] Adding an OPTIONAL _task-<label> to structural MRI acquisitions

### PR DESCRIPTION
**Pull request by melanieganz**
As was discussed in:

- #1177 

it is of interest to add the `OPTIONAL _task-<label>` to the structural MRI specifications to allow for task such as pre-described motion paradigms.

### CODE REVIEW

**Review by bendhouseart — APPROVED** [review]
I don't know what's wrong with the link checker, but I also don't care as it's very finicky. Approving per the discussion in issue #1177

File: src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md
Comment (TheChymera) [inline src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:322]: I would recommend removing the reference to the functional imaging, the label is what it is, without reference to other entries which may or may not always remain consistent.

Additionally, I think it should be highlighted that variations in acquisition quality (be they extraneous sources of noise or just head movement) are specifics of acquisition and not tasks in the context of a biological response (more information [here](https://github.com/bids-standard/bids-specification/issues/1177#issuecomment-1205817585)).

Perhaps:

"""
[...] can be used in order to document tasks performed during anatomical MR acquisition, the effects of which may pertain to brain anatomy. Note that variations in acquisition quality, as produced by either the subject or extraneous sources, are acquisition parameters and not tasks. Each task has a unique label that MUST only consist of letters and/or numbers (other characters, including spaces and underscores, are not allowed). Those labels MUST be consistent across subjects and sessions.
"""

**Review by TheChymera — CHANGES_REQUESTED** [review]
I think the scope of the task entity as defined here should be revised in light of the original use case which motivated its addition. While I agree that tasks can apply to anatomical scans, in the current formulation this is (1) too specific as it equates task with movement, and (2) too broad and may mutate the task entity into “everything that is moving while the experiment is happening”.

A clear delineation needs to be made between what is an acquisition parameter and what is a task, even if it is not the one I proposed.

**Review by Remi-Gau — COMMENTED** [review]
Actually... We are also gonna need some metadata added to this as well I think.

- TaskName
- TaskDescription

From the top of my head... No pun intended.

Maybe some others?

**Comment by Remi-Gau** [comment]
- "Instructions" seems to be another metadata field

From a quick browse into the other I don't see how the others task related field could make sense here (cog atlas ID, cog paradigm ontology ID...) but I may be wrong...

**Comment by melanieganz** [comment]
Dear all, 

thanks all for your feedback! I think we need to discuss this further before adapting it somehow and added some more questions back our [discussion](https://github.com/bids-standard/bids-specification/issues/1177#issuecomment-1205817585). Let me know what you think.

File: src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md
Comment (bendhouseart) [reply src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:322]: > can be used in order to document tasks performed during anatomical MR acquisition, the effects of which may pertain to brain anatomy. Note that variations in acquisition quality, as produced by either the subject or extraneous sources, are acquisition parameters and not tasks. 

I think this definition is too limiting. And I'm not sure we should be so quick to define things such as quality; those definitions should be left up to the researcher/study as we have no idea what they're measuring.

More broadly defining tasks to be say an action/inaction (task for lack of a better word) that is performed by the subject would be my preference. Also, using `task-` in this manner allows for the use of all the accompanying features already incorporated within, e.g. events files.  Which to me makes even more of an argument for broadening task outside of `fmri`. see [here](https://bids-specification.readthedocs.io/en/stable/04-modality-specific-files/05-task-events.html)

**Comment by Remi-Gau** [comment]
To move forward on this, I would suggest a  dividing and conquer approach.

Let's focus on adding in this PR "task" to "anat" without any example.

This should include the typical metadata usually associated with task.

@melanieganz i will a task list in the top message of this PR.

Then we can keep discussing the suggested example in the associated PR.

**Comment by melanieganz** [comment]
So you want to approve the above and merge? If you find more people that approve, I am all on board. :-)

**Comment by Remi-Gau** [comment]
> So you want to approve the above and merge? If you find more people that approve, I am all on board. :-)

OK to approve once [those metadata](https://github.com/bids-standard/bids-specification/pull/1185#issue-1329165100) are added

File: src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md
Comment (tsalo) [reply src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:322]: I agree that subject actions requested by the experimenter should count as a task (e.g., resting-state scans). I still think events files should be required when the `task` entity is present though.

Comment (Remi-Gau) [inline src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:322]:
```suggestion
The OPTIONAL [`task-<label>`](../99-appendices/09-entities.md#task) entity can be used in order to allow tasks during structural MR acquisitions, e.g. pre-described motion paradigms such as nodding, to be described. Task labels MUST be consistent across subjects and sessions.
```

Comment (Remi-Gau) [reply src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:322]: Remove label definition as it is already in the common principles.

**Comment by melanieganz** [comment]
Sorry everyone, I have been out travelling and then started teaching and tyhis is why this has been hanging. I am trying to resolve it this week. But I might ping @Remi-Gau and @bendhouseart to help with this PR GitHub mess. Since I am "all powerful" as steering group member in this repo I don't want to mess things up, because I am doing things wrong. ;-)

**Review by melanieganz — COMMENTED** [review]
Approve Remi's additions

**Comment by melanieganz** [comment]
Dear all, I now added the recommended metadata that @Remi-Gau  asked for:
- DONE TaskName
- DONE TaskDescription
- DONE Instructions 
I added it directly after the mentioning of task in anatomical imaging and I shortened the descriptions so no mentions of task are there. Are they too close to the section on 'Task (including resting state) imaging data'? 

I think there is a small issue in the macro for the TaskName remaining. It automatically states: For example "TaskName" "faces n-back" will correspond to task label facesnback. Maybe we can agree to take this out of the Macro and add it or make a more general formulation in the macro tables. I will recommend this change also.

**Comment by melanieganz** [comment]
I made a proposed change to the TaskName example in order to not make it only focused on functional MRI, let me know if that makes sense to you or not, see [here](https://github.com/melanieganz/bids-specification/blob/master/src/schema/objects/metadata.yaml).

File: src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md
Comment (Remi-Gau) [inline src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:322]:
```suggestion
The OPTIONAL [`task-<label>`](../99-appendices/09-entities.md#task) entity can be used
in order to allow tasks during structural MR acquisitions, 
for example pre-described motion paradigms such as nodding, to be described. 
Task labels MUST be consistent across subjects and sessions.
```

Comment (Remi-Gau) [reply src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:322]: just for line breaks and to remove latin abbreviation

Comment (Remi-Gau) [reply src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:322]: @effigies the "bad latin" workflow did not catch this case of abbreviation and it seems that the validation workflow is not running. 

Is this expected? Am I missing something?

File: src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md
Comment (Remi-Gau) [inline src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:213]:
```suggestion
The OPTIONAL [`task-<label>`](../99-appendices/09-entities.md#task) entity can be used
in order to allow tasks during structural MR acquisitions,
for example pre-described motion paradigms such as nodding, to be described.
Task labels MUST be consistent across subjects and sessions.
```

Comment (Remi-Gau) [reply src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:213]: markdown fix

File: src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md
Comment (Remi-Gau) [inline src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:230]:
```suggestion
{{ MACROS___make_metadata_table(
   {
      "TaskName": "RECOMMENDED",
      "TaskDescription": "RECOMMENDED",
      "Instructions": "RECOMMENDED",
   }
) }}
```

Comment (Remi-Gau) [reply src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:230]: this too should pacify the remark validation

File: src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md
Comment (Remi-Gau) [inline src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:210]:
```suggestion
The OPTIONAL [`task-<label>`](../99-appendices/entities.md#task) entity can be used
```

We have removed the number prefix for those files:

https://github.com/bids-standard/bids-specification/blob/master/src/appendices/entities.md

File: src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md
Comment (melanieganz) [reply src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:213]: Thanks for fixing all the formatting, @Remi-Gau!!!

File: src/schema/rules/datatypes/anat.yaml
Comment (Remi-Gau) [inline src/schema/rules/datatypes/anat.yaml:30]: This is triggering a validation error of the schema because entities should be in the right order in those files.

https://github.com/bids-standard/bids-specification/actions/runs/3188511667/jobs/5201262576

I have tried to make the warnings informative enough so that it would tell you what the right order is.

https://github.com/bids-standard/bids-specification/actions/runs/3188511667/jobs/5201262576#step:5:86

Long story short: `task` comes before `acquisition

```python
  suffix group nonparametric in anat is out of order:
                  - got: ['subject', 'session', 'acquisition', 'ceagent', 'reconstruction', 'run', 'part', 'task']
                  - should be: ['subject', 'session', 'task', 'acquisition', 'ceagent', 'reconstruction', 'run', 'part']
                  
    warnings.warn(

tools/schemacode/bidsschematools/tests/test_rules.py::test_entity_order
  /home/runner/work/bids-specification/bids-specification/tools/schemacode/bidsschematools/tests/test_rules.py:125: UserWarning: 
                  
  suffix group parametric in anat is out of order:
                  - got: ['subject', 'session', 'acquisition', 'ceagent', 'reconstruction', 'task', 'run']
                  - should be: ['subject', 'session', 'task', 'acquisition', 'ceagent', 'reconstruction', 'run']
                  
    warnings.warn(

tools/schemacode/bidsschematools/tests/test_rules.py::test_entity_order
  /home/runner/work/bids-specification/bids-specification/tools/schemacode/bidsschematools/tests/test_rules.py:125: UserWarning: 
                  
  suffix group defacemask in anat is out of order:
                  - got: ['subject', 'session', 'acquisition', 'ceagent', 'reconstruction', 'run', 'modality', 'task']
                  - should be: ['subject', 'session', 'task', 'acquisition', 'ceagent', 'reconstruction', 'run', 'modality']
                  
    warnings.warn(

tools/schemacode/bidsschematools/tests/test_rules.py::test_entity_order
  /home/runner/work/bids-specification/bids-specification/tools/schemacode/bidsschematools/tests/test_rules.py:125: UserWarning: 
                  
  suffix group multiecho in anat is out of order:
                  - got: ['subject', 'session', 'acquisition', 'ceagent', 'reconstruction', 'run', 'echo', 'part', 'task']
                  - should be: ['subject', 'session', 'task', 'acquisition', 'ceagent', 'reconstruction', 'run', 'echo', 'part']
                  
    warnings.warn(

tools/schemacode/bidsschematools/tests/test_rules.py::test_entity_order
  /home/runner/work/bids-specification/bids-specification/tools/schemacode/bidsschematools/tests/test_rules.py:125: UserWarning: 
                  
  suffix group multiflip in anat is out of order:
                  - got: ['subject', 'session', 'acquisition', 'ceagent', 'reconstruction', 'run', 'echo', 'flip', 'part', 'task']
                  - should be: ['subject', 'session', 'task', 'acquisition', 'ceagent', 'reconstruction', 'run', 'echo', 'flip', 'part']
                  
    warnings.warn(

tools/schemacode/bidsschematools/tests/test_rules.py::test_entity_order
  /home/runner/work/bids-specification/bids-specification/tools/schemacode/bidsschematools/tests/test_rules.py:125: UserWarning: 
                  
  suffix group multiinversion in anat is out of order:
                  - got: ['subject', 'session', 'acquisition', 'ceagent', 'reconstruction', 'run', 'inversion', 'part', 'task']
                  - should be: ['subject', 'session', 'task', 'acquisition', 'ceagent', 'reconstruction', 'run', 'inversion', 'part']
                  
    warnings.warn(

tools/schemacode/bidsschematools/tests/test_rules.py::test_entity_order
  /home/runner/work/bids-specification/bids-specification/tools/schemacode/bidsschematools/tests/test_rules.py:125: UserWarning: 
                  
  suffix group mp2rage in anat is out of order:
                  - got: ['subject', 'session', 'acquisition', 'ceagent', 'reconstruction', 'run', 'echo', 'flip', 'inversion', 'part', 'task']
                  - should be: ['subject', 'session', 'task', 'acquisition', 'ceagent', 'reconstruction', 'run', 'echo', 'flip', 'inversion', 'part']
                  
    warnings.warn(

tools/schemacode/bidsschematools/tests/test_rules.py::test_entity_order
  /home/runner/work/bids-specification/bids-specification/tools/schemacode/bidsschematools/tests/test_rules.py:125: UserWarning: 
                  
  suffix group vfamt in anat is out of order:
                  - got: ['subject', 'session', 'acquisition', 'ceagent', 'reconstruction', 'run', 'echo', 'flip', 'mtransfer', 'part', 'task']
                  - should be: ['subject', 'session', 'task', 'acquisition', 'ceagent', 'reconstruction', 'run', 'echo', 'flip', 'mtransfer', 'part']
                  
    warnings.warn(

tools/schemacode/bidsschematools/tests/test_rules.py::test_entity_order
  /home/runner/work/bids-specification/bids-specification/tools/schemacode/bidsschematools/tests/test_rules.py:125: UserWarning: 
                  
  suffix group mtr in anat is out of order:
                  - got: ['subject', 'session', 'acquisition', 'ceagent', 'reconstruction', 'run', 'mtransfer', 'part', 'task']
                  - should be: ['subject', 'session', 'task', 'acquisition', 'ceagent', 'reconstruction', 'run', 'mtransfer', 'part']               
```

File: src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md
Comment (melanieganz) [reply src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:210]: Ah this is probably a leftover from my fork of the spec not being up to date...let me also update it.

File: src/schema/rules/datatypes/anat.yaml
Comment (melanieganz) [reply src/schema/rules/datatypes/anat.yaml:30]: Ok, but wasn't this correctly done in the .yaml files before?

File: src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md
Comment (Remi-Gau) [inline src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:210]:
```suggestion
The OPTIONAL [`task-<label>`](../appendices/entities.md#task) entity can be used
```

Sorry. I was too quick. We also removed the prefix from the folder name, hence this doc building is still failing... 🙈

Comment (melanieganz) [reply src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:210]: Ok, maybe it's easier if I re-update my fork as well?

File: src/schema/rules/datatypes/anat.yaml
Comment (Remi-Gau) [reply src/schema/rules/datatypes/anat.yaml:30]: nope you had added task as the last entity in the list.

like this

```yaml
  entities:
    subject: required
    session: optional
    acquisition: optional
    ceagent: optional
    reconstruction: optional
    run: optional
    part: optional
    task: optional # not OK
```

and it should be like this


```yaml
  entities:
    subject: required
    session: optional
    task: optional # OK
    acquisition: optional
    ceagent: optional
    reconstruction: optional
    run: optional
    part: optional
```

make sense ?

File: src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md
Comment (Remi-Gau) [inline src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:229]:
```suggestion
) }}

```

pretty sure we need to skip a line for the table generation macro to not start using the next paragraph as input

@tsalo seen that happen before?

File: src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md
Comment (Remi-Gau) [reply src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md]: yup this may help too. 😄

**Comment by Remi-Gau** [comment]
> The table is looking funky. Let me see check this.
> 
> Screenshot from 2022-10-05 11-29-21

OK this got fixed by a line skip after the macro.

**Comment by Remi-Gau** [comment]
> * Are they too close to the section on 'Task (including resting state) imaging data'?

Not a problem for me. In case we reorganize things one day those things could then end up very far apart.

**Comment by Remi-Gau** [comment]
> Maybe we can agree to take this out of the Macro and add it or make a more general formulation in the macro tables. I will recommend this change also.

The change you have made makes sense to me.

File: src/schema/rules/datatypes/anat.yaml
Comment (effigies) [reply src/schema/rules/datatypes/anat.yaml:30]: I'm not sure why we enforce this, btw. The ordering here is not meant to be authoritative, and this is stated in the docs.

Comment (Remi-Gau) [reply src/schema/rules/datatypes/anat.yaml:30]: makes it easier for tools to know how to order entities in file names without having to refer to another part of the schema

File: src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md
Comment (Remi-Gau) [reply src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:213]: Actually @melanieganz given that you are linking to the definition in the appendices, you can remove the last sentence as it is already in the definition.

@satra I have always more understood as label `task-foo` for subject 1 must **mean** the same thing as `task-foo` for subject 2. I don't think that for any other type of task based things in BIDS we have put restrictions of the type  "all participants must have done all the tasks." But the fact that you are asking the question mean we should probably clarify that in the task definition.

File: src/schema/rules/datatypes/anat.yaml
Comment (melanieganz) [reply src/schema/rules/datatypes/anat.yaml:30]: Ok, then let me just reorder things

Comment (melanieganz) [reply src/schema/rules/datatypes/anat.yaml]: So the reordering is done for anat.

File: src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md
Comment (melanieganz) [reply src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:213]: Ok, then I will remove the last sentence here, but then we should make a separate PR or issue to clarify what is meant by "consistent" in the task description. I will do that as well and tag you.

Comment (melanieganz) [reply src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md]: Done.

**Review by Remi-Gau — APPROVED** [review]

File: src/schema/objects/metadata.yaml
Comment (melanieganz) [reply src/schema/objects/metadata.yaml:3142]: Good idea!

File: src/schema/rules/datatypes/anat.yaml
Comment (melanieganz) [reply src/schema/rules/datatypes/anat.yaml:25]: In my case, I will have motion tracking files available and will distribute those with it. But I am not sure how often this is available in general. Typically, people in the MoCo community give vocal instructions at a certain point of time in the seqiuence. So I am not sure this makes sense here.

File: src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md
Comment (melanieganz) [reply src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:228]: Ok, I will do that, thanks for making me aware of this

Comment (Remi-Gau) [reply src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md]: I had overlooked that too. I am used yet to "think with schema" yet.

File: src/schema/rules/datatypes/anat.yaml
Comment (melanieganz) [reply src/schema/rules/datatypes/anat.yaml:25]: @effigies are you happy with my answer?

Comment (melanieganz) [reply src/schema/rules/datatypes/anat.yaml:25]: ping @effigies ?

Comment (effigies) [reply src/schema/rules/datatypes/anat.yaml:25]: Yes, sorry for missing these.

**Review by TheChymera — COMMENTED** [review]
@melanieganz thanks for pinging me on this. Well... I still believe this fudges the issue, and that “tasks” during anatomical scans are basically acquisition confounds and should belong squarely under `acq-`. However, it seems basically everybody else is on board, and @bendhouseart additionally brought up a very good point here https://github.com/bids-standard/bids-specification/pull/1185#discussion_r939222589 namely that leveraging `task-` would help document this acquisition confound via the `_events.tsv` file feature. Given all of that I don't think it would make much sense for me to hold this back. Hopefully this won't lead to a proliferation of possible task values to encompass anything and everything that is happening during acquisition and time will prove my concerns to have been unfounded.

**Comment by effigies** [comment]
Alright. Two checkmarks and no blocks, let's merge.

**Pull request merged**
